desc/objectives/_generic.py

```diff
@@ -15,6 +19,262 @@
+            self.constants
+
+        Returns
+        -------
+        f : ndarray
+            Computed quantity.
+
+        """
+        f = self._fun_wrapped(params)
+        return f
+
+    def _jaxify(self, func, abstract_eval):
+        """Make an external (python) function work with JAX.
+
+        Positional arguments to func can be differentiated,
+        use keyword args for static values and non-differentiable stuff.
+
+        Note: Only forward mode differentiation is supported currently.
+
+        Parameters
+        ----------
+        func : callable
+            Function to wrap. Should be a "pure" function, in that it has no side
+            effects and doesn't maintain state. Does not need to be JAX transformable.
+        abstract_eval : callable
+            Auxilliary function that computes the output shape and dtype of func.
+            **Must be JAX transformable**. Should be of the form
+
+                abstract_eval(*args, **kwargs) -> Pytree with same shape and dtype as
+                func(*args, **kwargs)
+
+            For example, if func always returns a scalar:
+
+                abstract_eval = lambda *args, **kwargs: jnp.array(1.)
+
+            Or if func takes an array of shape(n) and returns a dict of arrays of
+            shape(n-2):
+
+                abstract_eval = lambda arr, **kwargs:
+                {"out1": jnp.empty(arr.size-2), "out2": jnp.empty(arr.size-2)}
+
+        Returns
+        -------
+        func : callable
+            New function that behaves as func but works with jit/vmap/jacfwd etc.
+
+        """
+        import jax
+
+        def wrap_pure_callback(func):
+            @functools.wraps(func)
+            def wrapper(*args, **kwargs):
+                result_shape_dtype = abstract_eval(*args, **kwargs)
+                return jax.pure_callback(
+                    func,
+                    result_shape_dtype,
+                    *args,
+                    vectorized=bool(self._vectorized),
+                    **kwargs,
+                )
+
+            return wrapper
+
+        def define_fd_jvp(func):
+            func = jax.custom_jvp(func)
+
+            @func.defjvp
+            def func_jvp(primals, tangents):
+                primal_out = func(*primals)
+
+                # flatten everything into 1D vectors for easier finite differences
+                y, unflaty = jax.flatten_util.ravel_pytree(primal_out)
+                x, unflatx = jax.flatten_util.ravel_pytree(primals)
+                v, _______ = jax.flatten_util.ravel_pytree(tangents)
+                # scale to unit norm if nonzero
+                normv = jnp.linalg.norm(v)
+                vh = jnp.where(normv == 0, v, v / normv)
+
+                def f(x):
+                    return jax.flatten_util.ravel_pytree(func(*unflatx(x)))[0]
+
+                tangent_out = (f(x + self._fd_step * vh) - y) / self._fd_step * normv
```

Comment:
we may want to also scale fd_step by `norm(x)` to ensure that dx << x

Reply:
normalize vh to be same norm as |x|

Reply:
(and undo at the end as well considering the scales)

Reply:
`x` here is really `params`, which includes variables at very different scales like boundary coefficients vs current profile coefficients. We are using the same finite difference step size for everything, which means that step could be huge for some variables and tiny for others. Then when generating wout files and running external codes for those perturbed equilibria, the results might not be physically meaningful.